### PR TITLE
Remove the usage of VLAs in the code

### PR DESCRIPTION
**Pull request by thufschmitt**
Although Variable-length arrays are allowed by the C standard, they are frowned
upon, and aren't valid C++ (`gcc` accepts them a [non-standard extension](https://gcc.gnu.org/onlinedocs/gcc/Variable-Length.html)) and `clang` apparently does the same which is why they work here). So better get rid of them altogether.

Depends on <https://github.com/NixOS/nix/pull/7346>.

### CODE REVIEW

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -1709,7 +1710,8 @@ void EvalState::callFunction(Value & fun, size_t nrArgs, Value * * args, Value &
                 /* We have all the arguments, so call the primop with
                    the previous and new arguments. */
 
-                Value * vArgs[arity];
+                assert(arity < 64);
```
Comment (cole-h) [inline src/libexpr/eval.cc:1713]: Should we maybe have a constant for this 64? I see it used a lot here.

That, or a comment explaining where it came from wherever it's used.

File: src/libexpr/primops.cc [hunk 2]
```diff
@@ -2470,8 +2469,8 @@ static void prim_catAttrs(EvalState & state, const PosIdx pos, Value * * args, V
     auto attrName = state.symbols.create(state.forceStringNoCtx(*args[0], pos));
     state.forceList(*args[1], pos);
 
-    Value * res[args[1]->listSize()];
-    unsigned int found = 0;
+    boost::container::small_vector<Value *, 256> res(args[1]->listSize());
```
Comment (cole-h) [inline src/libexpr/primops.cc:2472]: Same with this 256?

**Comment by edolstra** [comment]
I think statically sized arrays or `small_vector` should be avoided in functions that can be executed recursively. E.g.
```
                Value * vArgs[64];
```
in `callFunction()` adds 512 bytes of stack space to function calls, which adds up quickly for deeply recursive Nix code.

To be honest I don't really share the concerns over VLAs. We've used them for years without any issues. The zero-length VLA issue seems a bit language lawyering.

**Comment by tomberek** [comment]
TODO (from meeting):
- [ ] check eval impacts
- [ ] update stack overflow test

File: src/libexpr/eval.cc (hunk 1 above)
Comment (roberth) [reply src/libexpr/eval.cc]: Done. See `maxPrimOpArity`.

File: src/libexpr/primops.cc (hunk 2 above)
Comment (roberth) [reply src/libexpr/primops.cc:2731]: This is properly arbitrary. I think it's safe to assume that these numbers are guesswork, whether that's actual guesswork or an outdated benchmark.

**Comment by roberth** [comment]
> I think statically sized arrays or `small_vector` should be avoided in functions that can be executed recursively.

I've reduced the numbers significantly. Most significantly, 4 is enough for normal lambda calls. I expect the primops to contribute less to stack depth by number of nested invocations, so these use larger numbers, albeit reduced from what was originally in the PR.
This reduction is fine, because for large numbers the allocation is amortized to insignificance, and in many cases the data size is small anyway.



> TODO (from meeting):
> 
> * [ ]  check eval impacts

This is infinitely faster on expressions where these VLAs caused stack overflows.

> * [ ]  update stack overflow test

I find this somewhat pointless since the code does not allocate variable amounts of stack space anymore. Yes, tests are good, but these would be expensive, especially in terms of maintaining them because it involves developing non-trivial test support code for it. Silvan could attest that this is non-trivial. I'd rather focus that effort on more impactful testing improvements.

**Comment by roberth** [comment]
bug label: Fixes a stack overflow: https://github.com/NixOS/nixpkgs/pull/214021#discussion_r1394896936

File: src/libexpr/eval.cc [hunk 3]
```diff
@@ -1748,11 +1767,16 @@ void ExprCall::eval(EvalState & state, Env & env, Value & v)
     Value vFun;
     fun->eval(state, env, vFun);
 
-    Value * vArgs[args.size()];
+    // Nixpkgs lambdas by regex e.g. ([a-zA-Z]+:(\s|(/\*.*\/)|(#.*\n))*){5}
```
Comment (thufschmitt) [inline src/libexpr/eval.cc:1770]:
```suggestion
    // Empirical arity of Nixpkgs lambdas by regex e.g. ([a-zA-Z]+:(\s|(/\*.*\/)|(#.*\n))*){5}
```

(took me a few seconds to understand what that meant)

File: src/libexpr/eval.hh [hunk 4]
```diff
@@ -18,6 +18,12 @@
 
 namespace nix {
 
+/**
+ * We put a limit on primop arity because it lets us use a fixed size array on
+ * the stack. 16 is already an impractical number of arguments. Use an attrset
```
Comment (thufschmitt) [inline src/libexpr/eval.hh:23]:
```suggestion
 * the stack. 8 is already an impractical number of arguments. Use an attrset
```

I expect this should match the actual value

**Review by thufschmitt — COMMENTED** [review]
Thanks for taking over this @roberth !

Looks good to me, although I think this deserves some benchmarks (IIRC, the original version had a small impact depending on the sizes chosen)

**Comment by roberth** [comment]
> deserves some benchmarks

That would be great, but I can only treat this as a bugfix unfortunately.
Based on your earlier measurement I feel confident that this is not a performance regression.

File: src/libexpr/eval.cc (hunk 3 above)
Comment (roberth) [reply src/libexpr/eval.cc]: Thanks!

File: src/libexpr/primops.cc (hunk 2 above)
Comment (roberth) [reply src/libexpr/primops.cc]: I did name and document this after all. See `nonRecursiveStackReservation`

**Comment by thufschmitt** [comment]
> Based on your earlier measurement I feel confident that this is not a performance regression.

I wasn't, but I hyperfined it and there's zero visible difference on `nix search nixpkgs firefox`. So I should have :stuck_out_tongue:

**Comment by edolstra** [comment]
This broke the `buildNoGC` job: https://hydra.nixos.org/build/241067941 which is blocking the new release.

**Comment by roberth** [comment]
> This broke the `buildNoGC` job: https://hydra.nixos.org/build/241067941 which is blocking the new release.

- Fixed in #9374

**Pull request merged**